### DISCUSSION TITLE
Can azul be used in web?

### DISCUSSION THREAD

**Discussion by 6174** [maps4print/azul#307] [Ideas]
Hi @fschutt F, azul is really amazing .  I am just very curious about can azul be ported to web which use webgl?  The scenario is to create a cross platform canvas based app( Something like figma).   

I know azul is build for GUI scenario, but I think it's also can be used in drawing or creating scenario(UI Editing, Video editing etc). 

Thx :-)

**Comment by 6174** [comment]
Figma have very good performance. Which I think is the key differentiator compare to other products.  So I am trying to study which tech can achieve that.   As you can see normal web tech can't achieve that. I think azul maybe a solution

**Comment by fschutt** [comment]
No - but there are no fundamental problems to using Azul on the web. It's just not implemented. Azul already uses a DOM / CSS model - it should ideally not use WebGL, because that's useless if you can just use HTML / CSS. I.e.:

```rust
Dom::div()
.with_inline_style("width:10px; height:50px; background:red;")
```

... should translate to the web relatively well using [web-sys](https://rustwasm.github.io/wasm-bindgen/examples/dom.html). As for compilation - everything is compilable to wasm32, except for the `azul-desktop` crate (which does the shell integration).

In order to port azul to a new platform, you only have to implement the `App::get_monitors()` and `App::run()` methods. One feature that is missing however, is DOM diffing / reconciliation. Reconciliation is basically the core of any modern web framework and it's necessary for animations, too (i.e. imagine transitions between slides).

Right now I'm not interested in adding web support, as Azul was mainly made for CAD / GIS applications, i.e. applications where web browsers aren't performant enough.
